### ISSUE TITLE
Glean.initialize takes a long time on cold startup

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/fenix#7746]
This issue is to track issues upstream in the Glean SDK:
- https://bugzilla.mozilla.org/show_bug.cgi?id=1610574
- https://bugzilla.mozilla.org/show_bug.cgi?id=1610575

Assigning Alessio to make it clear the Glean team is working on it (but not because Alessio is working on it in particular).

The text below is from when the issue was previously filed and may not still be super relevant.

---

Related bugs:
- deprioritized consultation of Glean enabled/disabled performance: #6728
- regression in `Glean.initialize`: https://github.com/mozilla-mobile/fenix/issues/7617
- setStartupMetrics on worker thread: https://github.com/mozilla-mobile/fenix/issues/7281
- Glean initialization done separately: https://github.com/mozilla-mobile/fenix/issues/7386

---

Re-evaluation below refers to https://github.com/mozilla-mobile/fenix/issues/7746#issuecomment-574946550.

### Nimbledroid on Nexus 5, Jan 15 before re-evaluation
In investigating https://github.com/mozilla-mobile/fenix/issues/7580, I uploaded several `forPerformanceTest` builds to Nimbledroid with Glean disabled. Currently, we're at 1.74s on ND (aggregating the last 5 builds, which have a high variance from 1.79s to 1.69s). After removing Glean init (by commenting out GleanMetricsService initializer), we get 1.41s on ND over 3 runs (variance 1.36s to 1.44s): **that's a delta of 0.33s with Glean removed** or 18.97% improvement. These profiles were taken before Glean.initialize was moved as early as possible.

ND runs on a Nexus 5. In profiles, we've seen very long run times for Glean on the G5 (500ms+ in places). We haven't seen such drastic problems on the P2.

### Local testing on Pixel 2, Jan 19 after re-evaluation
I see a **9.14% improvement** to Displayed time (time to first frame drawn) on my Pixel 2 when removing glean (commenting out `Glean.initialize` and `GleanMetricsService`'s initializer): it goes from 395.83ms to 359.67ms: that's 36.16ms. I ran using geckobetaFenixBeta builds over 6 runs (not FNPRMS).

### Thoughts on addressing this issue
* Rope in the Glean team: they may have a good understanding about what we can change here
* Quick wins: let's eliminate any long running calls within Glean methods
  * P2 profiles indicate Glean.initialize, setStartupMetrics, and ReleaseMetricController.track (specifically for Glean) takes a long time but unclear how the debuggable flag may impact the results.
* More involved wins? (credit to Severin for thinking of this)
  * Why do Glean things need to run _during_ startup? Why can't they be deferred until visual completeness?
  * Why does _anything_ related to Glean have to run on the main thread?

**Comment by mcomella** [comment]
Spoke to the glean team: we should re-evaluate after https://github.com/mozilla-mobile/fenix/issues/7617 and the data migration code is removed https://github.com/mozilla-mobile/android-components/commit/843f3b774ba1ce88c5097574b2049d4898085baa (landed but needs to merge to fenix).

**Comment by Dexterp37** [comment]
> 
> 
> Spoke to the glean team: we should re-evaluate after #7617 and the data migration code is removed [mozilla-mobile/android-components@843f3b7](https://github.com/mozilla-mobile/android-components/commit/843f3b774ba1ce88c5097574b2049d4898085baa) (landed but needs to merge to fenix).

Note that this should be in today's Nightly.

**Comment by Dexterp37** [comment]
> After removing Glean init (by commenting out GleanMetricsService initializer),

Wait a second: could you instead just comment out `Glean.initialize()`?

**Comment by mcomella** [comment]
The changes have landed and Glean still has a very large impact on startup: in my re-evaluation, I saw Glean takes ~9.17% of startup time until first frame drawn. I wonder there may be new regressions introduced from moving `Glean.initialize` and changes in Glean SDK 24.

I added some generic thoughts on how we might want to approach this is in the first comment https://github.com/mozilla-mobile/fenix/issues/7746#issue-550525268.

@Dexterp37 FYI that we're using this to track a larger investigation into Glean performance. I'll have someone from the team reach out to you to see what we can do.

**Comment by mcomella** [comment]
> I saw Glean takes ~9.17% of startup time until first frame drawn

This seems disproportionate to the amount of effort I suspect is required for Glean but perhaps I don't fully understand all the work that's being done.

**Comment by Dexterp37** [comment]
> ### Local testing on Pixel 2, Jan 19 after re-evaluation
> 
> I see a **9.14% improvement** to Displayed time (time to first frame drawn) on my Pixel 2 when removing glean (commenting out `Glean.initialize` and `GleanMetricsService`'s initializer): it goes from 395.83ms to 359.67ms: that's 36.16ms. I ran using geckobetaFenixBeta builds over 6 runs (not FNPRMS).

Ok, I collected back my thoughts and I pieced back what I was trying to tell you the other day and then I forgot. You said:

> (commenting out `Glean.initialize` and `GleanMetricsService`'s initializer)

This is commenting out two different unrelated systems: `Glean.initialize` is part of the Glean SDK, `GleanMetricsService` is an abstraction layer added by the Fenix team to abstract away Glean. They might contribute differently to the startup performance, so I'd suggest to test by just commenting out `Glean.initialize`.

As far as I can tell, if you comment out `GleanMetricsService` initializer (which ends up calling [this](https://github.com/mozilla-mobile/fenix/blob/76bf74c4befa87b9f26bdfc3fafb7e927a8e53d6/app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt#L439)), it [early gets out of any collection call](https://github.com/mozilla-mobile/fenix/blob/76bf74c4befa87b9f26bdfc3fafb7e927a8e53d6/app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt#L453) that might happen even at startup. This means that you are probably not stumbling on [this](https://github.com/mozilla-mobile/fenix/blob/76bf74c4befa87b9f26bdfc3fafb7e927a8e53d6/app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt#L85) anymore, skipping almost entirely the Fenix abstraction of Glean.

> ### Thoughts on addressing this issue
> 
>     * Rope in the Glean team: they may have a good understanding about what we can change here

Please, reach out to me at any time! I'm happy to arrange a meeting ASAP and/or kick off the conversation in Berlin.

>     * Quick wins: let's eliminate any long running calls within Glean methods
>       
>       * P2 profiles indicate Glean.initialize, setStartupMetrics, and ReleaseMetricController.track (specifically for Glean) takes a long time but unclear how the debuggable flag may impact the results.

The only Glean SDK method mentioned above is `Glean.initialize`. Other methods are part of the Fenix abstraction made by the Fenix team. We could talk about that, but that's a Fenix team responsability down the line.

>     * More involved wins? (credit to Severin for thinking of this)
>       
>       * Why do Glean things need to run _during_ startup? Why can't they be deferred until visual completeness?

Yes, some Glean things are required to run _during_ startup. I'm happy to walk through our initialization process, as we did with @MarcLeclair, if needed.

>       * Why does _anything_ related to Glean have to run on the main thread?

In general we don't have such a requirement, nor we report that in our docs (which is our ultimate source of truth, minus bugs :-) ).
We only require the initialization method to be called on the main thread, as stated [in our docs](https://mozilla.github.io/glean/book/user/general-api.html#initializing-the-glean-sdk).

If there's any performance problem with `Glean.initialize`, we're happy to work with you folks and solve it for all products using the Glean SDK.
We are actively working on startup performance and listening to your feedback: see [this bug for flushing the startup queue off the main thread](https://bugzilla.mozilla.org/show_bug.cgi?id=1609052), which was just merged for Android (fix for iOS will come within this week).

**Comment by mcomella** [comment]
mleclair and I met with the Glean team: we gained a shared understanding of the current performance issues. Notably, there's a split between issues in the Glean SDK and issues with how Fenix uses Glean.

The Glean team filed two issues they plan to investigate wrt the former:
* https://bugzilla.mozilla.org/show_bug.cgi?id=1610574
* https://bugzilla.mozilla.org/show_bug.cgi?id=1610575

They noted they're not landing anything too risky until after All Hands, which includes the `flushQueuedTasks` change.

I'm going to repurpose this bug to track their work.

---

wrt the latter – how Fenix uses Glean, I filed https://github.com/mozilla-mobile/fenix/issues/7847 for that work.
